### PR TITLE
fix: refactor widget addition sagas and widget operation sagas to allow gridPosition in pasteSaga

### LINKED ISSUE #32962: Bug: Existing widgets are reshuffled when dragging a building block
This happens when there is a scroll on page, and adding multiple drag and drop BBs leads to existing widgets getting pushed down.

### PR DESCRIPTION
**Pull request by rahulbarwal**
**Problem**
The current building block on canvas implementation uses the pasteSaga logic, and this logic requires the mousePosition to know where to paste the widgets. Due to building blocks being mimicked as a single widget, we get grid position values instead of mouse location when dragging and dropping building blocks, this forced us to create e utility function that generates mousePosition from gridPosition values. This logic was flawed and led to building blocks placement being wrong when the view is scrolled.
**Solution**
We have decided to implement the pasteSaga to allow both mousePosition and gridPosition as expected values for dropping the copied widgets unto the canvas.

Fixes #32962

## Automation

/ok-to-test tags="@tag.Widget"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/8875384318>
> Commit: 6f425a394a7349b6461576b3762b0c152eed5659
> Cypress dashboard url: <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=8875384318&attempt=1" target="_blank">Click here!</a>

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [x] No

## Summary by CodeRabbit

- **New Features**
  - Enhanced building block functionalities including new actions and templates for a smoother drag-and-drop experience.
  - Added ability to specify exact grid positions when placing widgets, enhancing layout precision.

- **Refactor**
  - Streamlined sagas related to widget operations and building blocks for improved performance and maintainability.

- **Bug Fixes**
  - Adjusted widget addition logic to prevent UI issues during drag-and-drop actions.

- **Chores**
  - Removed outdated functions and reorganized imports to align with updated functionalities.

### CODE REVIEW

File: app/client/src/sagas/BuildingBlocksSagas.ts [hunk 1]
```diff
@@ -153,6 +179,220 @@ function* forkStarterBuildingBlockToApplicationSaga(
     yield call(saveCopiedWidgets, JSON.stringify(existingCopiedWidgets));
   }
 }
+
+function* addBuildingBlockActionsToApp(dragDetails: DragDetails) {
```
Comment (rahulbarwal) [inline app/client/src/sagas/BuildingBlocksSagas.ts:183]: @jacquesikot There should be a better name to this. The way I see it, this actually calls the API to /block api. Lets show that in the name.

File: app/client/src/sagas/BuildingBlocksSagas.ts [hunk 2]
```diff
@@ -153,6 +179,220 @@ function* forkStarterBuildingBlockToApplicationSaga(
...
+        isShortcut: false,
+      },
+    });
+    yield put({
+      type: ReduxActionErrorTypes.DRAGGING_BUILDING_BLOCK_TO_CANVAS_ERROR,
+    });
+  }
+}
+
+export function* addBuildingBlockToCanvasSaga(
```
Comment (rahulbarwal) [inline app/client/src/sagas/BuildingBlocksSagas.ts:354]: @jacquesikot Better name here as well?
here we are initiating the skeleton loading and then adding it to application.

File: app/client/src/sagas/BuildingBlocksSagas.ts [hunk 3]
```diff
@@ -153,6 +179,220 @@ function* forkStarterBuildingBlockToApplicationSaga(
...
+          disallowUndo: true,
+          isShortcut: false,
+        },
+      });
+
+      yield put({
+        type: ReduxActionTypes.PASTE_COPIED_WIDGET_INIT,
+        payload: {
+          groupWidgets: false,
+          gridPosition: {
```
Comment (rahulbarwal) [inline app/client/src/sagas/BuildingBlocksSagas.ts:286]: @jacquesikot I need final nod on this name :), I think we can do better.

**Review by rahulbarwal — COMMENTED** [review]
@jacquesikot My review revolves around nomenclature. I feel all sagas in the new file have similar name and will confuse us in future.

**Comment by mohanarpit** [comment]
@rahulbarwal @jacquesikot Do you think we can write a Jest unit test to assert that this problem won't show up again?

**Comment by rahulbarwal** [comment]
Hey @mohanarpit We have a separate PR for automation in works, We will ensure this case is added there as well.

**Comment by mohanarpit** [comment]
> Hey @mohanarpit We have a separate PR for automation in works, We will ensure this case is added there as well.

Can you please link to that PR? Also any reason we are writing tests in a separate PR from the actual code? This way of working sounds like a violation of our coding best practices.

**Review by jacquesikot — APPROVED** [review]

File: app/client/src/sagas/WidgetAdditionSagas.ts
[the remark this answers is not in the document]
Comment (rahulbarwal) [reply app/client/src/sagas/WidgetAdditionSagas.ts:1]: Done.

**Comment by jacquesikot** [comment]
@mohanarpit PR for cypress tests for this feature #33036

**Comment by mohanarpit** [comment]
> @mohanarpit PR for cypress tests for this feature #33036

@jacquesikot Thanks. But my question still stands. Why is the PR for Cypress test separate from the change itself? Why are we not writing the code changes and Cypress tests in the same PR? This way, there is a changelog for our future selves about the code change & the test case to ensure that the code change doesn't regress.

**Pull request merged**